types/object.go

```diff
@@ -0,0 +1,603 @@
+package types
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/NeowayLabs/abad/internal/utf16"
+)
+
+type (
+	// DataObject is a collection of named values.
+	DataObject struct {
+		// Class is the kind of object
+		class         string
+		notExtensible bool
+		props         map[string]*PropertyDescriptor
+	}
+
+	callable interface {
+		Call(this *DataObject, args []Value) Value
+	}
+)
+
+var (
+	// S is an alias to make utf-16 strings.
+	S            = utf16.S
+	valueAttr    = S("value")
+	writableAttr = S("writable")
+	getAttr      = S("get")
+	setAttr      = S("set")
+	enumAttr     = S("enumerable")
+	cfgAttr      = S("configurable")
+
+	protoAttr    = S("prototype")
+	toStringAttr = S("toString")
+	valueOfAttr  = S("valueOf")
+)
+
+// DefaultPrototypeDesc is a base prototype object that extends Null.
+// The root of the prototype-based type hierarchy.
+func DefaultPrototypeDesc() *PropertyDescriptor {
+	return NewDataPropDesc(Null, true, false, false)
+}
+
+// NewDataObject creates a new DataObject using proto as
+// prototype.
+func NewDataObject(proto Value) *DataObject {
+	obj := NewBaseDataObject()
+
+	// obj must extend proto
+	delete(obj.props, "prototype")
+
+	// error ignored because it does not fail if
+	// there's no previous properties.
+	ok, err := obj.DefineOwnPropertyP(protoAttr,
+		NewDataPropDesc(proto, false, true, false), true)
+
+	if !ok {
+		// should never occurs
+		panic(err)
+	}
+
+	return obj
+}
+
+// NewBaseDataObject is the same as ecmascript code:
+//   Object.create(null);
+// This is the root of the prototype chain.
+func NewBaseDataObject() *DataObject {
+	return NewDataObjectP(DefaultPrototypeDesc())
+}
+
+// NewDataObjectP creates an object with PropertyDescriptor
+// proto as prototype attribute.
+// The *everything is an object* concept makes
+// impossible to defineOwnProperties work by
+// passing an object to define an object (recursive
+// definition).
+func NewDataObjectP(proto *PropertyDescriptor) *DataObject {
+	obj := &DataObject{
+		class: "object",
+		props: make(map[string]*PropertyDescriptor),
+	}
+
+	obj.props["prototype"] = proto
```

Comment:
Actually you could already put the prototype key on the map above if you initialize it sintatically.